File: ICPC_Practice/Vietnam-Southern-ICPC-Contest-2025-main/code/ProA.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

long long max_of_three(long long a, long long b, long long c)
{
    return std::max({a, b, c});
}

void solve()
{
    int n;
    long long C, D;
    std::cin >> n >> C >> D;

    std::vector<long long> a(n), b(n);
    long long max_a = LLONG_MIN;
    long long max_b = LLONG_MIN;

    for (int i = 0; i < n; ++i)
    {
        std::cin >> a[i];
        if (a[i] > max_a)
        {
            max_a = a[i];
        }
    }
    for (int i = 0; i < n; ++i)
    {
        std::cin >> b[i];
        if (b[i] > max_b)
        {
            max_b = b[i];
        }
    }

    std::vector<long long> d(n);
    long long min_d = LLONG_MAX;
    long long max_d = LLONG_MIN;
    for (int i = 0; i < n; ++i)
    {
        d[i] = a[i] - b[i];
        if (d[i] < min_d)
        {
            min_d = d[i];
        }
        if (d[i] > max_d)
        {
            max_d = d[i];
        }
    }

    if (max_d - min_d > C + D)
    {
        std::cout << -1 << std::endl;
        return;
    }

    long long x_bound = max_a - C;
    long long y_bound = max_b - D;
    long long z_lower = max_d - C;
    long long z_upper = min_d + D;

    std::vector<std::pair<long long, long long>> candidates;

    if (z_lower <= x_bound - y_bound && x_bound - y_bound <= z_upper)
    {
        candidates.push_back({x_bound, y_bound});
    }

    long long y2 = x_bound - z_upper;
    if (y2 >= y_bound)
    {
        candidates.push_back({x_bound, y2});
    }
    long long y3 = x_bound - z_lower;
    if (y3 >= y_bound)
    {
        candidates.push_back({x_bound, y3});
    }

    long long x4 = y_bound + z_upper;
    if (x4 >= x_bound)
    {
        candidates.push_back({x4, y_bound});
    }
    long long x5 = y_bound + z_lower;
    if (x5 >= x_bound)
    {
        candidates.push_back({x5, y_bound});
    }

    long long min_ops = LLONG_MAX;

    if (candidates.empty())
    {
        std::cout << -1 << std::endl;
        return;
    }

    for (const auto &p : candidates)
    {
        long long x = p.first;
        long long y = p.second;
        long long current_ops = 0;
        for (int i = 0; i < n; ++i)
        {
            current_ops += max_of_three(0LL, x - a[i], y - b[i]);
        }
        min_ops = std::min(min_ops, current_ops);
    }

    std::cout << min_ops << std::endl;
}

int main()
{
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    int T;
    std::cin >> T;
    while (T--)
    {
        solve();
    }

    return 0;
}

File: ICPC_Practice/Vietnam-Southern-ICPC-Contest-2025-main/code/ProA.py
import sys

def solve():
    """
    Hàm giải quyết cho một test case.
    """
    try:
        # Đọc dữ liệu đầu vào
        line1 = sys.stdin.readline()
        if not line1: return
        n, C, D = map(int, line1.split())
        a = list(map(int, sys.stdin.readline().split()))
        b = list(map(int, sys.stdin.readline().split()))
    except (IOError, ValueError):
        return

    # --- Bước 1: Tính các đại lượng cần thiết ---
    
    # Tính mảng hiệu d[i] = a[i] - b[i], đây là đại lượng bất biến
    d = [a[i] - b[i] for i in range(n)]

    # Tìm các giá trị max/min ban đầu
    max_a = max(a)
    max_b = max(b)
    min_d = min(d)
    max_d = max(d)

    # --- Bước 2: Kiểm tra điều kiện tồn tại nghiệm ---
    
    # Nếu chênh lệch lớn nhất của hiệu (max_d - min_d) lớn hơn
    # tổng các khoảng cho phép (C + D), thì không thể tìm ra lời giải.
    if max_d - min_d > C + D:
        print(-1)
        return

    # --- Bước 3: Xác định không gian tìm kiếm và các điểm ứng cử viên ---
    
    # Gọi min_a' và min_b' là min của mảng a và b sau khi biến đổi.
    # Ta cần tìm cặp (min_a', min_b') tối ưu.
    # Các điều kiện ràng buộc tạo thành một vùng đa giác lồi.
    # Chi phí là hàm lồi, nên giá trị tối ưu sẽ nằm ở các đỉnh của vùng này.
    
    # Ràng buộc dưới cho min_a' và min_b'
    x_bound = max_a - C  # min_a' >= max_a - C
    y_bound = max_b - D  # min_b' >= max_b - D

    # Ràng buộc cho hiệu min_a' - min_b'
    # z_lower <= min_a' - min_b' <= z_upper
    z_lower = max_d - C
    z_upper = min_d + D

    # Tìm các đỉnh của vùng khả thi làm ứng cử viên
    candidates = []
    
    # Điểm giao của các đường biên. Chúng ta chỉ quan tâm các điểm
    # nằm trên biên "dưới-trái" của vùng khả thi.
    
    # Ứng cử viên 1: giao của x = x_bound và y = y_bound
    if z_lower <= x_bound - y_bound <= z_upper:
        candidates.append((x_bound, y_bound))
    
    # Ứng cử viên 2 & 3: giao của x = x_bound và các đường chéo
    y2 = x_bound - z_upper
    if y2 >= y_bound:
        candidates.append((x_bound, y2))
        
    y3 = x_bound - z_lower
    if y3 >= y_bound:
        candidates.append((x_bound, y3))

    # Ứng cử viên 4 & 5: giao của y = y_bound và các đường chéo
    x4 = y_bound + z_upper
    if x4 >= x_bound:
        candidates.append((x4, y_bound))
        
    x5 = y_bound + z_lower
    if x5 >= x_bound:
        candidates.append((x5, y_bound))
        
    # --- Bước 4: Tính chi phí và tìm kết quả nhỏ nhất ---
    
    min_ops = float('inf')

    # Nếu không có ứng cử viên nào, có thể có lỗi logic hoặc trường hợp đặc biệt.
    # Tuy nhiên, với logic trên, danh sách ứng cử viên sẽ không rỗng nếu có nghiệm.
    if not candidates:
        print(-1) # Trường hợp dự phòng
        return

    # Duyệt qua các điểm ứng cử viên để tìm chi phí nhỏ nhất
    for x, y in candidates:
        current_ops = 0
        for i in range(n):
            # Số phép toán tại chỉ số i là max(0, min_a' - a[i], min_b' - b[i])
            ops_i = max(0, x - a[i], y - b[i])
            current_ops += ops_i
        min_ops = min(min_ops, current_ops)

    print(min_ops)


def main():
    """
    Hàm chính để đọc số lượng test case và gọi hàm giải.
    """
    try:
        # Đọc số lượng test case
        T_str = sys.stdin.readline()
        if not T_str: return
        T = int(T_str)
        for _ in range(T):
            solve()
    except (IOError, ValueError):
        return

if __name__ == "__main__":
    main()